### ISSUE TITLE
Shouldn't Array::pop_front() and pop_back() return pop-ed element?

### ISSUE THREAD

**Issue by StraToN** [godotengine/godot#3763]
It seems that pop methods from Array only resize the array, but don't return the actual pop-ed element. What's the point of these methods if they just call Array::resize() ?

https://github.com/godotengine/godot/blob/master/core/array.cpp#L275

**Comment by volzhs** [comment]
+1

**Comment by akien-mga** [comment]
I guess it wouldn't hurt to return the pop-ed element instead of just void indeed.

**Comment by punto-** [comment]
do we have a back() method? it always awkward to do array[array.size()-1]
when using them as stacks

**Comment by akien-mga** [comment]
In the STL it looks like those functions only remove the first and last elements without returning it though:
- http://www.cplusplus.com/reference/list/list/pop_front/
- http://www.cplusplus.com/reference/list/list/pop_back/

However in Python the pop() function does return the pop-ed element: http://www.tutorialspoint.com/python/list_pop.htm

**Comment by akien-mga** [comment]
> do we have a back() method? it always awkward to do array[array.size()-1]
when using them as stacks

Doesn't look like it, but it could easily be added I guess if the ``back()`` method is common in other languages. Another alternative would be to add support for ``array[-1]``, e.g. change:
```
Variant& Array::operator[](int p_idx) {

	return _p->array[p_idx];
}
```
for:
```
Variant& Array::operator[](int p_idx) {

	if (p_idx < 0) {
		return _p->array[_p->array.size() - p_idx];
	} else {
		return _p->array[p_idx];
	}
}
```
(untested)

**Comment by StraToN** [comment]
Potential reason: Exception safety.
Source: http://blogs.msdn.com/b/zhanli/archive/2010/06/29/c-tips-why-the-pop-method-of-stl-stack-does-not-return-a-value.aspx

Explanation: 
Suppose pop() methods do return the pop-ed element.
Now, let the client write this:
```
Variant v = array.pop_front(); // will call Variant copy constructor
```
or
```
Variant v;
v = array.pop_front(); // will call Variant operator=
```

What *may* happen is that either Variant's copy constructor or Variant's operator= throw an exception: the pop-ed element was not correctly inserted in variable v, although the array did change, since it was resized in the pop() function.

Thus, one has to first get the value we want, and then only call pop().
But why doesn't Python bother?

**Comment by punto-** [comment]
we don't use exceptions tho

**Comment by pkowal1982** [comment]
I vote for returning those poped values instead of void

**Comment by StraToN** [comment]
@pkowal1982 Read https://github.com/godotengine/godot/issues/3763#issuecomment-186131805 above. 

TL;DR: quoting myself:
> So, a good solution would simply consist in:
>     creating new methods Array::back() and Array::top() that get the element we are supposed to get, respectively first and last ones.
>     proposing new methods pop_front_element() and pop_back_element() that actually copy the element obtained with front() or back() before actually call pop_front() or pop_back(), and finally return the copied element.

Unless anyone has another idea? @punto- @akien-mga @bojidar-bg @akien-mga

**Comment by vnen** [comment]
What's the difference between adding a `pop_front_element()` and making `pop_front()` return the element? Also "copy" in this context mean an actual duplicate of the object (which would be likely not intended)?

In any case, `front()` and `back()` are good additions IMO.

**Comment by StraToN** [comment]
@vnen Code talks more than words :)

```c++
void Array::pop_front() {
    // doesn't change
    // resizes the Array without returning the front element
}

Variant Array::pop_front_element() {
    Variant element  = this->get_front();
    this->pop_front();
    return element;
}

Variant Array::front() const {
    return this->get(0);
}```

**Comment by pkowal1982** [comment]
I think having `pop_front()` and `pop_front_element()` is too much. `pop_front()` should return poped element and if you don't want it just ignore it...
`front()` and `back()` are OK for me.

**Comment by StraToN** [comment]
It's not that I want it or not. Adding either solution is trivial.
The reason I didn't do it it is that I don't want to implement something that potentially introduces bugs or non-optimized code. Pushing forward for a solution or the other doesn't interest me if there is not argumentation with it.

**Comment by kbake** [comment]
I was looking into this and it seems that it's been addressed, may be good to close. Current (v3.0.alpha.custom_build) implementations below:

    Variant Array::front() const {
	ERR_FAIL_COND_V(_p->array.size() == 0, Variant());
	return operator[](0);
    }

    Variant Array::back() const {
	ERR_FAIL_COND_V(_p->array.size() == 0, Variant());
	return operator[](_p->array.size() - 1);
    }

...

    Variant Array::pop_back(){
	if (!_p->array.empty()) {
		int n = _p->array.size() - 1;
		Variant ret = _p->array.get(n);
		_p->array.resize(n);
		return ret;
	}
	return Variant();
    }

    Variant Array::pop_front(){
	if (!_p->array.empty()) {
		Variant ret = _p->array.get(0);
		_p->array.remove(0);
		return ret;
	}
	return Variant();
    }

Only thing of concern is the comment made on the commit for the `pop_`'s: 

> pop_front, pop_back now return values instead of `void`. Things I didn't know how to properly implement:
    1. pop_front & pop_back shows in the help menu Object as return value. I know this is incorrect but if not Object than what? Cause it can't be void. It needs to be a generic type that includes all the Array types

**Comment by akien-mga** [comment]
> pop_front, pop_back now return values instead of void. Things I didn't know how to properly implement:
1. pop_front & pop_back shows in the help menu Object as return value. I know this is incorrect but if not Object than what? Cause it can't be void. It needs to be a generic type that includes all the Array types

The `BIND_METHOD` call needs to use `pop_front:Variant` and `pop_back:Variant` so that the type is enforced in the docs.

**Comment by akien-mga** [comment]
Closing as fixed by #6304.
